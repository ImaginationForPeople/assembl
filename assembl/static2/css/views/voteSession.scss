.votesession-page {
  h1,
  h2,
  h3 {
    margin-bottom: 0;
  }

  .vote-instructions {
    padding: 0 10px;
    font-size: $font-size-m;
    font-family: $font-family4;
    max-width: 780px;
    line-height: 32px;
  }

  .content-section {
    padding: 0;
  }

  .available-tokens {
    margin: 20px 0 40px;

    .category-available-tokens {
      display: flex;
      padding: 0 0 15px;
      max-width: 780px;
      margin-left: -5px;

      h2 {
        margin-top: 7px;
      }

      .tokens {
        padding-left: 100px;
        padding-top: 9px;
      }

      .text {
        margin: 0;
      }

      span {
        font-size: $font-size-xs;
      }
    }
  }

  .available-tokens-sticky {
    position: fixed;
    top: 60px;
    left: 0;
    right: 0;
    z-index: 100;
    border-top: 1px solid $grey1;
    box-shadow: 3px 4px 7px $grey1;

    .separator {
      display: none;
    }

    .category-available-tokens {
      .tokens {
        padding: 20px 0 0 0;
        display: inline-block;
        vertical-align: top;
        text-align: left;

        svg {
          margin: 0 3px;
        }
      }

      .text {
        line-height: 20px;
        margin: 0 0 5px;
        display: inline-block;
        text-align: left;

        .dark-title-2 {
          font-size: $font-size-s;
          line-height: 18px;
        }

        span {
          font-size: $font-size-xs;
        }
      }
    }
  }

  .extra-margin-top {
    margin-top: 400px;
  }

  .tokens > * {
    margin: 7px 3px;
  }

  .theme-box {
    &:first-child {
      margin: -70px 0 0;
    }
  }

  .proposal.row {
    margin: 90px 10px 70px;

    .proposal-title {
      font-family: $font-family;
      margin-top: 0;
      font-size: 22px;
    }

    .text {
      line-height: 24px;
    }

    .tokens-line {
      margin-bottom: 0;

      p {
        margin: 0;
      }

      .tokens {
        margin-left: -5px;
      }
    }

    .tokens > button {
      border: none;
      padding: 0;
      background: none;

      &:focus {
        outline: none;
      }

      &:active {
        box-shadow: none;
      }
    }

    .tokens > .btn-default:hover {
      background-color: transparent;
    }
  }

  .form-actions {
    margin-top: 50px;
    margin-bottom: 100px;
  }

  .link-button {
    border: none;
    font-size: $font-size-xs;
    color: $link-color;
    text-decoration: none;
    background-color: $grey0;
    padding: 15px;

    &:hover {
      color: $link-hover-color;
      text-decoration: none;
    }

    &:focus {
      outline: none;
    }

    &:active {
      box-shadow: none;
    }

    [class^='assembl-icon-']:before,
    [class*=' assembl-icon-']:before {
      font-size: $font-size-xxs;
      margin-left: 5px;
    }
  }

  .link-button:focus,
  .link-button:active,
  .link-button:hover {
    background-color: transparent;
  }

  .expand-row {
    margin: 0 40px;
  }

  .participants-count {
    .assembl-icon-profil:before {
      font-size: 40px;
      margin-bottom: 8px;
    }
  }

  .participants-count__text {
    color: $first-color;
    padding: 0 10px 20px 10px;
  }

  .proposal-results {
    .participants-count {
      border: 1px solid $grey5;
      margin: 20px 0;
      min-height: 120px;
    }

    .doughnut-container {
      text-align: center;
      height: 150px;
      width: 150px;
      padding: 10px 0 0;
      position: relative;
      margin: auto;

      .doughnut {
        .circle {
          stroke-width: 18px;
        }
      }

      .doughnut-totalCount {
        position: absolute;
        top: 63px;
        left: 53px;
        width: 45px;
        font-size: $font-size-s;

        .vote-totalCount {
          font-size: $font-size-sm;
          font-weight: bold;
          font-family: $font-family3;
        }
      }
    }
  }

  .votes-in-progress {
    margin: 0 40px;
    padding: 0 15px 15px;

    .padding-s {
      padding-left: 5px;
      padding-right: 5px;
    }

    .box {
      border: 1px solid $grey5;
      margin-top: 10px;
      height: 140px;
      text-align: center;
    }

    .tokens-box {
      text-align: center;
    }

    .doughnut-container {
      margin: 0 auto 0 auto;
      height: 80px;
      width: 80px;
      position: relative;

      .doughnut {
        .circle {
          stroke-width: 18px;
        }
      }

      .doughnut-totalCount {
        position: absolute;
        top: 23px;
        left: 24px;
        width: 34px;
        font-size: $font-size-xxs;

        .vote-totalCount {
          font-size: $font-size-xs;
          font-family: $font-family3;
        }
      }
    }
  }

  .expand-result {
    padding-bottom: 20px;
  }

  .gauge-box {
    padding: 20px 0 0;
  }

  .gauge-box__instructions {
    font-size: $font-size-s;
  }

  .gauge-box__estimate {
    color: $first-color;
  }

  .separator {
    border-bottom: 1px $btn-dark-border-style $grey5;
    max-width: 605px;
    margin-left: -5px;
  }
}

.token-tooltip {
  border-radius: 4px;

  .tooltip-inner {
    background: none;
  }
}

@media screen and (min-width: $screen-sm-min) {
  .votesession-page {
    .proposal.row {
      padding: 15px 5px 15px 0;
      display: flex;
    }

    .participants-count {
      height: inherit;
      padding: 20px 0 0;
    }

    .tokens-box {
      padding: 19px 0 19px 0;
      text-align: center;
    }
  }
}

@media screen and (max-width: $screen-xs-max) {
  .votesession-page {
    .votes-in-progress {
      padding: 0 0 15px 0;

      .box {
        margin-top: 15px;
        height: 160px;
      }
    }

    .available-tokens-sticky {
      display: none;
    }

    .container {
      padding: 0;
      margin: 0;
    }

    .proposal.row {
      padding: 15px 0 15px;
    }

    .participants-count {
      padding: 32px 0 0;
    }

    .tokens-box {
      padding: 10px 0 20px 0;
      text-align: center;
    }

    .section-title {
      display: block;
    }

    .max-container {
      padding-left: 15px;
      padding-right: 15px;
    }
  }
}

@media screen and (min-width: $screen-xl-min) {
  .proposal-results {
    width: 80%;
  }
}

@media screen and (max-width: 660px) {
  .votesession-page {
    .available-tokens {
      .category-available-tokens {
        display: block;

        .tokens {
          padding-left: 10px;
          margin-left: -5px;
        }
      }
    }
  }
}

@media screen and (max-width: 420px) {
  .votesession-page {
    .title-section .dark-title-1 {
      margin: 0 5px;
    }
  }
}
