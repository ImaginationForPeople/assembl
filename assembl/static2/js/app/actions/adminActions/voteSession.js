// @flow
import * as actionTypes from '../actionTypes';
import { createRandomId } from '../../utils/globalFunctions';

export const updateVoteSessionPageTitle = (locale: string, value: string): actionTypes.UpdateVoteSessionPageTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_TITLE
});

export const updateVoteSessionPageSeeCurrentVotes = (value: string): actionTypes.UpdateVoteSessionPageSeeCurrentVotes => ({
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_SEECURRENTVOTES
});

export const updateVoteSessionPageSubtitle = (locale: string, value: string): actionTypes.UpdateVoteSessionPageSubtitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_SUBTITLE
});

export const updateVoteSessionPageInstructionsTitle = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPageInstructionsTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_INSTRUCTIONS_TITLE
});

export const updateVoteSessionPageInstructionsContent = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPageInstructionsContent => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_INSTRUCTIONS_CONTENT
});

export const updateVoteSessionPagePropositionsTitle = (
  locale: string,
  value: string
): actionTypes.UpdateVoteSessionPagePropositionsTitle => ({
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_PROPOSITIONS_TITLE
});

export const updateVoteSessionHeaderImage = (value: File): actionTypes.UpdateVoteSessionHeaderImage => ({
  value: value,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE_IMAGE
});

export const updateVoteModules = (voteModules: actionTypes.VoteModulesArray): actionTypes.UpdateVoteModules => ({
  voteModules: voteModules,
  type: actionTypes.UPDATE_VOTE_MODULES
});

export const createTokenVoteModule = (id: string): actionTypes.CreateTokenVoteModule => ({
  id: id,
  type: actionTypes.CREATE_TOKEN_VOTE_MODULE
});

export const deleteVoteModule = (id: string): actionTypes.DeleteVoteModule => ({
  id: id,
  type: actionTypes.DELETE_VOTE_MODULE
});

export const createGaugeVoteModule = (id: string): actionTypes.CreateGaugeVoteModule => ({
  id: id,
  type: actionTypes.CREATE_GAUGE_VOTE_MODULE
});

export const updateTokenVoteExclusiveCategory = (id: string, value: boolean): actionTypes.UpdateTokenVoteExclusiveCategory => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_EXCLUSIVE_CATEGORY
});

export const updateTokenVoteInstructions = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateTokenVoteInstructions => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_INSTRUCTIONS
});

export const createTokenVoteCategory = (id: string, moduleId: string): actionTypes.CreateTokenVoteCategory => ({
  id: id,
  moduleId: moduleId,
  type: actionTypes.CREATE_TOKEN_VOTE_CATEGORY
});

export const deleteTokenVoteCategory = (moduleId: string, index: number): actionTypes.DeleteTokenVoteCategory => ({
  moduleId: moduleId,
  index: index,
  type: actionTypes.DELETE_TOKEN_VOTE_CATEGORY
});

export const updateTokenVoteCategoryTitle = (
  id: string,
  locale: string,
  value: string,
  moduleId: string
): actionTypes.UpdateTokenVoteCategoryTitle => ({
  moduleId: moduleId,
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_CATEGORY_TITLE
});

export const updateTokenTotalNumber = (id: string, value: number, moduleId: string): actionTypes.UpdateTokenTotalNumber => ({
  moduleId: moduleId,
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_TOTAL_NUMBER
});

export const updateTokenVoteCategoryColor = (
  id: string,
  value: string,
  moduleId: string
): actionTypes.UpdateTokenVoteCategoryColor => ({
  moduleId: moduleId,
  id: id,
  value: value,
  type: actionTypes.UPDATE_TOKEN_VOTE_CATEGORY_COLOR
});

export const updateGaugeVoteInstructions = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateGaugeVoteInstructions => ({
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_INSTRUCTIONS
});

export const updateGaugeVoteIsNumber = (id: string, value: boolean): actionTypes.UpdateGaugeVoteIsNumber => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_IS_NUMBER
});

export const updateGaugeVoteNbTicks = (id: string, value: number): actionTypes.UpdateGaugeVoteNbTicks => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_NUMBER_TICKS
});

export const createGaugeVoteChoice = (moduleId: string, id: string): actionTypes.CreateGaugeVoteChoice => ({
  moduleId: moduleId,
  id: id,
  type: actionTypes.CREATE_GAUGE_VOTE_CHOICE
});

export const deleteGaugeVoteChoice = (moduleId: string, index: number): actionTypes.DeleteGaugeVoteChoice => ({
  moduleId: moduleId,
  index: index,
  type: actionTypes.DELETE_GAUGE_VOTE_CHOICE
});

export const updateGaugeVoteChoiceLabel = (
  id: string,
  locale: string,
  value: string,
  moduleId: string
): actionTypes.UpdateGaugeVoteChoiceLabel => ({
  moduleId: moduleId,
  id: id,
  locale: locale,
  value: value,
  type: actionTypes.UPDATE_GAUGE_VOTE_CHOICE_LABEL
});

export const updateGaugeMinimum = (id: string, value: number): actionTypes.UpdateGaugeMinimum => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_MINIMUM
});

export const updateGaugeMaximum = (id: string, value: number): actionTypes.UpdateGaugeMaximum => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_MAXIMUM
});

export const updateGaugeUnit = (id: string, value: string): actionTypes.UpdateGaugeUnit => ({
  id: id,
  value: value,
  type: actionTypes.UPDATE_GAUGE_UNIT
});

export const updateVoteSessionPage = (value: Object): actionTypes.UpdateVoteSessionPage => ({
  id: value.id,
  titleEntries: value.titleEntries,
  seeCurrentVotes: value.seeCurrentVotes,
  subTitleEntries: value.subTitleEntries,
  instructionsSectionTitleEntries: value.instructionsSectionTitleEntries,
  instructionsSectionContentEntries: value.instructionsSectionContentEntries,
  propositionsSectionTitleEntries: value.propositionsSectionTitleEntries,
  headerImage: value.headerImage,
  type: actionTypes.UPDATE_VOTE_SESSION_PAGE
});

export const updateVoteProposals = (voteProposals: actionTypes.VoteProposalsArray): actionTypes.UpdateVoteProposals => ({
  voteProposals: voteProposals,
  type: actionTypes.UPDATE_VOTE_PROPOSALS
});

export const createVoteProposal = (id: string): actionTypes.CreateVoteProposal => ({
  id: id,
  type: actionTypes.CREATE_VOTE_PROPOSAL
});

export const deleteVoteProposal = (id: string): actionTypes.DeleteVoteProposal => ({
  id: id,
  type: actionTypes.DELETE_VOTE_PROPOSAL
});

export const updateVoteProposalTitle = (id: string, locale: string, value: string): actionTypes.UpdateVoteProposalTitle => ({
  id: id,
  value: value,
  locale: locale,
  type: actionTypes.UPDATE_VOTE_PROPOSAL_TITLE
});

export const updateVoteProposalDescription = (
  id: string,
  locale: string,
  value: string
): actionTypes.UpdateVoteProposalDescription => ({
  id: id,
  value: value,
  locale: locale,
  type: actionTypes.UPDATE_VOTE_PROPOSAL_DESCRIPTION
});

export const moveProposalUp = (id: string): actionTypes.MoveProposalUp => ({
  id: id,
  type: actionTypes.MOVE_PROPOSAL_UP
});

export const moveProposalDown = (id: string): actionTypes.MoveProposalDown => ({
  id: id,
  type: actionTypes.MOVE_PROPOSAL_DOWN
});

export const addModuleToProposal = (
  id: string,
  proposalId: string,
  voteSpecTemplateId: string
): actionTypes.AddModuleToProposal => ({
  id: id,
  proposalId: proposalId,
  voteSpecTemplateId: voteSpecTemplateId,
  type: actionTypes.ADD_MODULE_TO_PROPOSAL
});

export const createVoteProposalAndModules = (id: string) => (dispatch: Function, getState: Function) => {
  dispatch(createVoteProposal(id));
  const { modulesInOrder } = getState().admin.voteSession;
  modulesInOrder.forEach(moduleId => dispatch(addModuleToProposal(createRandomId(), id, moduleId)));
};

export const undeleteModule = (id: string): actionTypes.UndeleteModule => ({
  id: id,
  type: actionTypes.UNDELETE_MODULE
});

/* if a vote spec template params changes, set _hasChanged to true on all its dependencies */
export const markAllDependenciesAsChanged = (id: string): actionTypes.MarkAllDependenciesAsChanged => ({
  id: id,
  type: actionTypes.MARK_ALL_DEPENDENCIES_AS_CHANGED
});

export const setValidationErrors = (id: string, errors: ValidationErrors): actionTypes.SetValidationErrors => ({
  errors: errors,
  id: id,
  type: actionTypes.SET_VALIDATION_ERRORS
});

export const cancelModuleCustomization = (id: string): actionTypes.CancelModuleCustomization => ({
  id: id,
  type: actionTypes.CANCEL_MODULE_CUSTOMIZATION
});