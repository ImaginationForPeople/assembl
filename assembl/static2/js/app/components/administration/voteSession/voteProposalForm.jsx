// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n, Translate } from 'react-redux-i18n';
import { OverlayTrigger, Button, Checkbox, FormGroup, HelpBlock } from 'react-bootstrap';
import { type RawContentState } from 'draft-js';

import FormControlWithLabel from '../../common/formControlWithLabel';
import { getEntryValueForLocale } from '../../../utils/i18n';
import { deleteVoteProposalTooltip, upTooltip, downTooltip } from '../../common/tooltips';
import {
  updateVoteProposalTitle,
  updateVoteProposalDescription,
  deleteVoteProposal,
  moveProposalUp,
  moveProposalDown,
  addModuleToProposal,
  cancelModuleCustomization,
  deleteVoteModule,
  undeleteModule
} from '../../../actions/adminActions/voteSession';
import { displayCustomModal, displayModal, closeModal } from '../../../utils/utilityManager';
import { createRandomId } from '../../../utils/globalFunctions';
import CustomizeGaugeForm from './customizeGaugeForm';

type VoteProposalFormProps = {
  index: number,
  title: string,
  description: RawContentState,
  _toDelete: boolean,
  markAsToDelete: Function,
  updateTitle: Function,
  updateDescription: Function,
  editLocale: string,
  nbProposals: number,
  handleUpClick: Function,
  handleDownClick: Function,
  associateModuleToProposal: Function,
  cancelCustomization: Function,
  deassociateModuleToProposal: Function,
  reactivateModule: Function,
  tokenModules: Object,
  gaugeModules: Object,
  proposalModules: Object,
  refetchVoteSession: Function,
  validationErrors: ValidationErrors
};

export const getValidationState = (validationErrors: ?Array<ErrorDef>): ?string =>
  (validationErrors && validationErrors.length > 0 ? 'error' : null);

const DumbVoteProposalForm = ({
  index,
  title,
  description,
  _toDelete,
  markAsToDelete,
  updateTitle,
  updateDescription,
  editLocale,
  nbProposals,
  handleUpClick,
  handleDownClick,
  tokenModules,
  gaugeModules,
  proposalModules,
  associateModuleToProposal,
  cancelCustomization,
  deassociateModuleToProposal,
  reactivateModule,
  refetchVoteSession,
  validationErrors
}: VoteProposalFormProps) => {
  if (_toDelete) {
    return null;
  }

  const handleTitleChange = e => updateTitle(editLocale, e.target.value);
  const handleDescriptionChange = value => updateDescription(editLocale, value);

  const moduleIsSelected = voteSpecTemplateId =>
    proposalModules.some(m => m.get('voteSpecTemplateId') === voteSpecTemplateId && !m.get('_toDelete'));

  const toggleModule = (voteSpecTemplateId) => {
    const pModule = proposalModules.find(m => m.get('voteSpecTemplateId') === voteSpecTemplateId);
    if (pModule) {
      if (pModule.get('_toDelete')) {
        reactivateModule(pModule.get('id'));
      } else {
        deassociateModuleToProposal(pModule.get('id'));
      }
    } else {
      associateModuleToProposal(voteSpecTemplateId);
    }
  };

  const confirmModal = () => {
    const modalTitle = <Translate value="administration.voteProposals.deleteModalTitle" />;
    const body = <Translate value="administration.voteProposals.deleteModalBody" />;
    const footer = [
      <Button key="cancel" onClick={closeModal} className="button-cancel button-dark">
        <Translate value="debate.confirmDeletionButtonCancel" />
      </Button>,
      <Button key="delete" onClick={markAsToDelete} className="button-submit button-dark">
        <Translate value="debate.confirmDeletionButtonDelete" />
      </Button>
    ];
    const includeFooter = true;
    return displayModal(modalTitle, body, includeFooter, footer);
  };

  const settingsModal = (id) => {
    const content = (
      <CustomizeGaugeForm close={closeModal} gaugeModuleId={id} editLocale={editLocale} refetchVoteSession={refetchVoteSession} />
    );
    displayCustomModal(content, true, 'gauge-settings-modal');
  };

  return (
    <div className="form-container vote-proposal-form">
      <div className="pointer right">
        <div className="inline">
          {index < nbProposals ? (
            <OverlayTrigger placement="top" overlay={downTooltip}>
              <Button onClick={handleDownClick} className="admin-icons">
                <span className="assembl-icon-down-bold grey" />
              </Button>
            </OverlayTrigger>
          ) : null}
          {index > 1 ? (
            <OverlayTrigger placement="top" overlay={upTooltip}>
              <Button onClick={handleUpClick} className="admin-icons">
                <span className="assembl-icon-up-bold grey" />
              </Button>
            </OverlayTrigger>
          ) : null}
        </div>
        {nbProposals > 2 && (
          <OverlayTrigger placement="top" overlay={deleteVoteProposalTooltip}>
            <Button className="admin-icons" onClick={confirmModal}>
              <span className="assembl-icon-delete grey" />
            </Button>
          </OverlayTrigger>
        )}
      </div>

      <div className="title">
        <Translate value="administration.voteProposals.defineProposal" number={index} />
      </div>

      <FormControlWithLabel
        value={title}
        label={I18n.t('administration.voteProposals.title')}
        onChange={handleTitleChange}
        required
        type="text"
        validationErrors={validationErrors.title}
      />

      <FormControlWithLabel
        key={`description-${editLocale}`}
        value={description}
        label={I18n.t('administration.voteProposals.description')}
        onChange={handleDescriptionChange}
        type="rich-text"
        required
      />

      <FormGroup validationState={getValidationState(validationErrors.modules)}>
        {validationErrors.modules &&
          validationErrors.modules.length > 0 &&
          validationErrors.modules.map((error, idx) => (
            <HelpBlock key={idx}>
              <Translate value={`administration.voteProposals.validationErrors.${error.code}`} {...error.vars} />
            </HelpBlock>
          ))}

        {tokenModules.map(voteSpecTemplateId => (
          <Checkbox
            key={voteSpecTemplateId}
            checked={moduleIsSelected(voteSpecTemplateId)}
            onChange={() => toggleModule(voteSpecTemplateId)}
          >
            <Translate value="administration.voteProposals.tokenVote" />
          </Checkbox>
        ))}
        {gaugeModules.map((voteSpecTemplateId, idx) => {
          const number = gaugeModules.size > 1 ? idx + 1 : '';
          const pModule = proposalModules.find(m => m.get('voteSpecTemplateId') === voteSpecTemplateId);
          return (
            <div key={voteSpecTemplateId}>
              <Checkbox
                className="inline"
                checked={moduleIsSelected(voteSpecTemplateId)}
                onChange={() => toggleModule(voteSpecTemplateId)}
              >
                {pModule && pModule.get('isCustom') ? (
                  <Translate value="administration.voteProposals.customGauge" number={number} />
                ) : (
                  <Translate value="administration.voteProposals.gauge" number={number} />
                )}
              </Checkbox>

              {pModule &&
                pModule.get('id') && (
                  <div>
                    <span
                      className="inline settings-link"
                      onClick={() => {
                        settingsModal(pModule.get('id'));
                      }}
                    >
                      <i className="assembl-icon-edit" />
                      <Translate value="administration.voteProposals.gaugeSettings" />
                    </span>

                    {pModule.get('isCustom') && (
                      <span
                        className="inline settings-link"
                        onClick={() => {
                          cancelCustomization(pModule.get('id'));
                        }}
                      >
                        <i className="assembl-icon-cancel" />
                        <Translate value="administration.voteProposals.cancelCustomization" />
                      </span>
                    )}
                  </div>
                )}
            </div>
          );
        })}
      </FormGroup>
      <div className="separator" />
    </div>
  );
};

const mapStateToProps = ({ admin }, { id, editLocale }) => {
  const proposal = admin.voteSession.voteProposalsById.get(id);
  const { modulesInOrder, modulesById } = admin.voteSession;
  const description = getEntryValueForLocale(proposal.get('descriptionEntries'), editLocale);
  return {
    _toDelete: proposal.get('_toDelete', false),
    validationErrors: proposal.get('_validationErrors'),
    title: getEntryValueForLocale(proposal.get('titleEntries'), editLocale),
    description: description && typeof description !== 'string' ? description.toJS() : null,
    order: proposal.get('order'),
    proposalModules: proposal.get('modules').map(moduleId => modulesById.get(moduleId)),
    tokenModules: modulesInOrder.filter(
      voteSpecTemplateId =>
        !modulesById.getIn([voteSpecTemplateId, 'proposalId']) &&
        modulesById.getIn([voteSpecTemplateId, 'type']) === 'tokens' &&
        !modulesById.getIn([voteSpecTemplateId, '_toDelete'])
    ),
    gaugeModules: modulesInOrder.filter(
      voteSpecTemplateId =>
        !modulesById.getIn([voteSpecTemplateId, 'proposalId']) &&
        modulesById.getIn([voteSpecTemplateId, 'type']) === 'gauge' &&
        !modulesById.getIn([voteSpecTemplateId, '_toDelete'])
    )
  };
};

const mapDispatchToProps = (dispatch, { id }) => ({
  cancelCustomization: moduleId => dispatch(cancelModuleCustomization(moduleId)),
  markAsToDelete: () => {
    dispatch(deleteVoteProposal(id));
    closeModal();
  },
  updateTitle: (locale, value) => {
    dispatch(updateVoteProposalTitle(id, locale, value));
  },
  updateDescription: (locale, value) => {
    dispatch(updateVoteProposalDescription(id, locale, value));
  },
  handleUpClick: () => dispatch(moveProposalUp(id)),
  handleDownClick: () => dispatch(moveProposalDown(id)),
  associateModuleToProposal: (voteSpecTemplateId) => {
    const newId = createRandomId();
    dispatch(addModuleToProposal(newId, id, voteSpecTemplateId));
  },
  deassociateModuleToProposal: moduleId => dispatch(deleteVoteModule(moduleId)),
  reactivateModule: moduleId => dispatch(undeleteModule(moduleId))
});

export { DumbVoteProposalForm };

export default connect(mapStateToProps, mapDispatchToProps)(DumbVoteProposalForm);