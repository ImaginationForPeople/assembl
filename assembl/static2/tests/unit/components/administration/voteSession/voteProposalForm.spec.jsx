import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { List } from 'immutable';

import { DumbVoteProposalForm } from '../../../../../js/app/components/administration/voteSession/voteProposalForm';
import '../../../../helpers/setupTranslations';

describe('VoteProposalForm component', () => {
  const markAsToDelete = jest.fn(() => {});
  const updateTitle = jest.fn(() => {});
  const updateDescription = jest.fn(() => {});
  const handleUpClick = jest.fn(() => {});
  const handleDownClick = jest.fn(() => {});

  it('should render an empty form to create a vote proposal without any checkbox below', () => {
    const props = {
      index: 1,
      title: null,
      description: null,
      _toDelete: false,
      markAsToDelete: markAsToDelete,
      updateTitle: updateTitle,
      updateDescription: updateDescription,
      editLocale: 'fr',
      nbProposals: 1,
      handleUpClick: handleUpClick,
      handleDownClick: handleDownClick,
      tokenModules: List(),
      gaugeModules: List(),
      validationErrors: {}
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbVoteProposalForm {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('should render an empty form to create a vote proposal with validation errors', () => {
    const props = {
      index: 1,
      title: null,
      description: null,
      _toDelete: false,
      markAsToDelete: markAsToDelete,
      updateTitle: updateTitle,
      updateDescription: updateDescription,
      editLocale: 'fr',
      nbProposals: 1,
      handleUpClick: handleUpClick,
      handleDownClick: handleDownClick,
      tokenModules: List(),
      gaugeModules: List(),
      validationErrors: {
        title: [
          {
            code: 'error.required',
            vars: {}
          }
        ]
      }
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbVoteProposalForm {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});