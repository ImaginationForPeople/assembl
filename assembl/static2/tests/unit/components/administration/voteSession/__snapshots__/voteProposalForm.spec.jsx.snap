// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteProposalForm component should render an empty form to create a vote proposal with validation errors 1`] = `
<div
  className="form-container vote-proposal-form"
>
  <div
    className="pointer right"
  >
    <div
      className="inline"
    >
      
    </div>
  </div>
  <div
    className="title"
  >
    <Translate
      number={1}
      value="administration.voteProposals.defineProposal"
    />
  </div>
  <FormControlWithLabel
    helperUrl=""
    label="Title of the proposal"
    labelAlwaysVisible={false}
    onChange={[Function]}
    required={true}
    type="text"
    validationErrors={
      Array [
        Object {
          "code": "error.required",
          "vars": Object {},
        },
      ]
    }
    value={null}
  />
  <FormControlWithLabel
    helperUrl=""
    label="Description"
    labelAlwaysVisible={false}
    onChange={[Function]}
    required={true}
    type="rich-text"
    validationErrors={null}
    value={null}
  />
  <FormGroup
    bsClass="form-group"
    validationState={null}
  >
    Immutable.List [
    ]
    Immutable.List [
    ]
  </FormGroup>
  <div
    className="separator"
  />
</div>
`;

exports[`VoteProposalForm component should render an empty form to create a vote proposal without any checkbox below 1`] = `
<div
  className="form-container vote-proposal-form"
>
  <div
    className="pointer right"
  >
    <div
      className="inline"
    >
      
    </div>
  </div>
  <div
    className="title"
  >
    <Translate
      number={1}
      value="administration.voteProposals.defineProposal"
    />
  </div>
  <FormControlWithLabel
    helperUrl=""
    label="Title of the proposal"
    labelAlwaysVisible={false}
    onChange={[Function]}
    required={true}
    type="text"
    validationErrors={null}
    value={null}
  />
  <FormControlWithLabel
    helperUrl=""
    label="Description"
    labelAlwaysVisible={false}
    onChange={[Function]}
    required={true}
    type="rich-text"
    validationErrors={null}
    value={null}
  />
  <FormGroup
    bsClass="form-group"
    validationState={null}
  >
    Immutable.List [
    ]
    Immutable.List [
    ]
  </FormGroup>
  <div
    className="separator"
  />
</div>
`;
